File: solver/cross_solver.py
"""
solver/cross_solver.py
=======================
Giai Cross (4 canh cua mat D) toi uu (it nuoc nhat) bang Pattern Database
o muc do "cubie" (vi tri + huong cua tung canh trong 12 khe), theo dung ky
thuat PDB kinh dien (Korf). Xem giai thich chi tiet trong edge_model.py.

Vi PDB cho khoang cach CHINH XAC trong khong gian rut gon nay, tu bat ky
trang thai nao (khac 0) luon ton tai it nhat 1 nuoc lam khoang cach giam
dung 1 -> chi can leo doc tham lam (greedy) la ra loi giai toi uu, khong
can A*/IDA*.
"""

from cube_engine import ALL_MOVES
from .edge_model import apply_move_to_edges, edges_from_state
from .pdb_builder import build_pdb_edges

TRACKED = ['DF', 'DB', 'DL', 'DR']   # 4 canh cross
_PDB = None


def get_pdb(verbose=False):
    global _PDB
    if _PDB is None:
        _PDB = build_pdb_edges('cross_edges', TRACKED, verbose=verbose)
    return _PDB


def solve_cross(state, max_moves=20):
    """Tra ve list nuoc di (Singmaster) giai Cross toi uu tu state hien tai.
    Khong thay doi state truyen vao."""
    pdb = get_pdb()
    pos, ori = edges_from_state(state, TRACKED)
    key = (tuple(pos), tuple(ori))
    d = pdb.get(key)
    if d is None:
        raise ValueError('Trang thai cross khong hop le / khong nam trong PDB')

    moves = []
    guard = 0
    while d > 0 and guard < max_moves:
        guard += 1
        for mv in ALL_MOVES:
            new_pos, new_ori = apply_move_to_edges(pos, ori, mv)
            k = (tuple(new_pos), tuple(new_ori))
            nd = pdb.get(k)
            if nd is not None and nd == d - 1:
                pos, ori = new_pos, new_ori
                moves.append(mv)
                d = nd
                break
        else:
            raise RuntimeError('Khong tim duoc nuoc giam khoang cach (bug PDB?)')
    return moves

File: rubik-core/src/cross_solver.rs
//! cross_solver.rs — port of solver/cross_solver.py
//! PDB cho khoảng cách CHÍNH XÁC -> từ bất kỳ trạng thái nào luôn có ít
//! nhất 1 nước làm khoảng cách giảm đúng 1, nên chỉ cần leo dốc tham lam.

use crate::cube::{CubeState, ALL_MOVES};
use crate::edge_model as EM;
use crate::pdb::{build_cross_pdb, CrossPdb};
use std::sync::OnceLock;

pub const TRACKED: [usize; 4] = [4, 5, 6, 7]; // DF, DB, DL, DR

static PDB: OnceLock<CrossPdb> = OnceLock::new();

pub fn get_pdb() -> &'static CrossPdb {
    PDB.get_or_init(|| build_cross_pdb(&TRACKED))
}

/// Trả về chuỗi nước Singmaster giải Cross tối ưu từ state hiện tại.
/// Không thay đổi state truyền vào.
pub fn solve_cross(state: &CubeState, max_moves: usize) -> Vec<&'static str> {
    let pdb = get_pdb();
    let (pos_v, ori_v) = EM::edges_from_state(state, &TRACKED);
    let mut pos: [usize; 4] = [pos_v[0], pos_v[1], pos_v[2], pos_v[3]];
    let mut ori: [u8; 4] = [ori_v[0], ori_v[1], ori_v[2], ori_v[3]];
    let mut d = pdb.get(&pos, &ori).expect("trạng thái cross không hợp lệ / không nằm trong PDB");

    let mut moves = Vec::new();
    let mut guard = 0;
    while d > 0 && guard < max_moves {
        guard += 1;
        let mut advanced = false;
        for &mv in ALL_MOVES.iter() {
            let (np, no) = EM::apply_move_to_edges(&pos, &ori, mv);
            let np4: [usize; 4] = [np[0], np[1], np[2], np[3]];
            let no4: [u8; 4] = [no[0], no[1], no[2], no[3]];
            if let Some(nd) = pdb.get(&np4, &no4) {
                if nd == d - 1 {
                    pos = np4;
                    ori = no4;
                    moves.push(mv);
                    d = nd;
                    advanced = true;
                    break;
                }
            }
        }
        if !advanced {
            panic!("Không tìm được nước giảm khoảng cách (bug PDB?)");
        }
    }
    moves
}
